File: Herencia/Ejercicio3/Persona.py
from datetime import datetime

class Persona:
    def __init__(self, ci, nombre, apellido, celular, fecha_Nac):
        self.ci = ci
        self.nombre = nombre
        self.apellido = apellido
        self.celular = celular
        self.fecha_Nac = datetime.strptime(fecha_Nac, "%Y-%m-%d")
        
    def edad(self):
        today = datetime.today()
        return today.year - self.fecha_Nac.year - ((today.month, today.day) < (self.fecha_Nac.month, self.fecha_Nac.day))
    
    def __str__(self):
        return f"{self.nombre} {self.apellido}, CI: {self.ci}, Celular: {self.celular}, Fecha Nac: {self.fecha_Nac.date()}"

class Estudiante(Persona):
    def __init__(self, ci, nombre, apellido, celular, fecha_Nac, ru, fecha_Ingreso, semestre):
        super().__init__(ci, nombre, apellido, celular, fecha_Nac)
        self.ru = ru
        self.fecha_Ingreso = fecha_Ingreso
        self.semestre = semestre

    def __str__(self):
        return f"Estudiante: {super().__str__()}, RU: {self.ru}, Fecha Ingreso: {self.fecha_Ingreso}, Semestre: {self.semestre}"

class Docente(Persona):
    def __init__(self, ci, nombre, apellido, celular, fecha_Nac, nit, profesion, especialidad, sexo):
        super().__init__(ci, nombre, apellido, celular, fecha_Nac)
        self.nit = nit
        self.profesion = profesion
        self.especialidad = especialidad
        self.sexo = sexo

    def __str__(self):
        return f"Docente: {super().__str__()}, NIT: {self.nit}, Profesión: {self.profesion}, Especialidad: {self.especialidad}, Sexo: {self.sexo}"

def main():
    # b) Implementa las clases con sus constructores, datos por defecto y mostrar
    estudiantes = [
        Estudiante("123", "Ana", "López", "70012345", "2000-05-12", "RU001", "2019-02-01", "6"),
        Estudiante("124", "Luis", "Pérez", "70054321", "1996-08-21", "RU002", "2017-02-01", "10"),
        Estudiante("125", "María", "López", "70088888", "2005-10-10", "RU003", "2023-02-01", "2")
    ]

    docentes = [
        Docente("300", "Carlos", "Gómez", "70123456", "1970-03-05", "NIT001", "Ingeniero", "Sistemas", "M"),
        Docente("301", "Sofía", "López", "70999999", "1980-07-20", "NIT002", "Licenciada", "Matemáticas", "F"),
        Docente("302", "Juan", "Pérez", "70222222", "1965-01-15", "NIT003", "Ingeniero", "Electrónica", "M")
    ]

    print("Lista de estudiantes:")
    for est in estudiantes:
        print(est)

    print("\nLista de docentes:")
    for doc in docentes:
        print(doc)

    # c) Mostrar los estudiantes mayores de 25 años
    print("\nEstudiantes mayores de 25 años:")
    for est in estudiantes:
        if est.edad() > 25:
            print(est)

    # d) Mostrar al docente que es “Ingeniero”, del sexo masculino y es el mayor de todos
    mayor_ingeniero = None
    for doc in docentes:
        if doc.profesion == "Ingeniero" and doc.sexo == "M":
            if mayor_ingeniero is None or doc.edad() > mayor_ingeniero.edad():
                mayor_ingeniero = doc

    print("\nDocente ingeniero masculino mayor:")
    if mayor_ingeniero:
        print(mayor_ingeniero)

    # e) Mostrar a los estudiantes y docentes que tienen el mismo apellido
    print("\nPersonas con apellidos coincidentes:")
    for est in estudiantes:
        for doc in docentes:
            if est.apellido == doc.apellido:
                print(f"{est.nombre} {est.apellido} (Estudiante) y {doc.nombre} {doc.apellido} (Docente)")

if __name__ == "__main__":
    main()

File: Herencia/Ejercicio3/Persona.java
import java.time.LocalDate;
import java.time.Period;
import java.util.ArrayList;
import java.util.List;

public class Persona {
    public String ci;
    public String nombre;
    public String apellido;
    public String celular;
    public LocalDate fechaNac;

    // b) CImplementa las clases con sus constructores, datos por defecto y mostrar
    public Persona(String ci, String nombre, String apellido, String celular, String fechaNac) {
        this.ci = ci;
        this.nombre = nombre;
        this.apellido = apellido;
        this.celular = celular;
        this.fechaNac = LocalDate.parse(fechaNac);
    }

    public int getEdad() {
        return Period.between(fechaNac, LocalDate.now()).getYears();
    }

    @Override
    public String toString() {
        return nombre + " " + apellido + ", CI: " + ci + ", Celular: " + celular + ", Fecha Nac: " + fechaNac;
    }
    
    public static void main(String[] args) {
        
        List<Estudiante> estudiantes = new ArrayList<>();
        estudiantes.add(new Estudiante("123", "Ana", "Lopez", "70012345", "2000-05-12", "RU001", "2019-02-01", "6"));
        estudiantes.add(new Estudiante("124", "Luis", "Perez", "70054321", "1996-08-21", "RU002", "2017-02-01", "10"));
        estudiantes.add(new Estudiante("125", "Maria", "Lopez", "70088888", "2005-10-10", "RU003", "2023-02-01", "2"));

        List<Docente> docentes = new ArrayList<>();
        docentes.add(new Docente("300", "Carlos", "Gomez", "70123456", "1970-03-05", "NIT001", "Ingeniero", "Sistemas", "M"));
        docentes.add(new Docente("301", "Sofia", "Lopez", "70999999", "1980-07-20", "NIT002", "Licenciada", "Matematicas", "F"));
        docentes.add(new Docente("302", "Juan", "Perez", "70222222", "1965-01-15", "NIT003", "Ingeniero", "Electronica", "M"));

        System.out.println("Lista de estudiantes:");
        for (Estudiante e : estudiantes) {
            System.out.println(e);
        }

        System.out.println("\nLista de docentes:");
        for (Docente d : docentes) {
            System.out.println(d);
        }

        // c) Mostrar los estudiantes mayores de 25 años
        System.out.println("\nEstudiantes mayores de 25 años:");
        for (Estudiante e : estudiantes) {
            if (e.getEdad() > 25) {
                System.out.println(e);
            }
        }

        // d) Mostrar al docente que tiene la profesión de “Ingeniero”, es del sexo masculino y es el mayor de todos
        Docente mayorIngeniero = null;
        for (Docente d : docentes) {
            if (d.getProfesion().equalsIgnoreCase("Ingeniero") && d.getSexo().equalsIgnoreCase("M")) {
                if (mayorIngeniero == null || d.getEdad() > mayorIngeniero.getEdad()) {
                    mayorIngeniero = d;
                }
            }
        }

        System.out.println("\nDocente ingeniero masculino mayor:");
        if (mayorIngeniero != null) {
            System.out.println(mayorIngeniero);
        }

        // e) Mostrar a los estudiantes y docentes que tienen el mismo apellido
        System.out.println("\nCoincidencias de apellido entre estudiantes y docentes:");
        for (Estudiante e : estudiantes) {
            for (Docente d : docentes) {
                if (e.getApellido().equalsIgnoreCase(d.getApellido())) {
                    System.out.println(e.nombre + " " + e.apellido + " (Estudiante) y " + d.nombre + " " + d.apellido + " (Docente)");
                }
            }
        }
    }
    
}
